File: moreapps/src/main/java/com/rocky/moreapps/utils/MoreAppsPrefHelper.java
package com.rocky.moreapps.utils;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;

import java.util.Set;
import java.util.TreeSet;

class MoreAppsPrefHelper {
    private static final String SHARED_PREFS_NAME = "MORE_APPS";

    private static MoreAppsPrefHelper instance;

    private SharedPreferences sharedPreferences;

    private MoreAppsPrefHelper(@NonNull Context context) {
        instance = this;
        sharedPreferences = context.getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE);
    }

    public static synchronized MoreAppsPrefHelper getInstance(@NonNull Context context) {
        if (instance == null) {
            instance = new MoreAppsPrefHelper(context);
        }

        return instance;
    }

    public void delete(String key) {
        if (sharedPreferences.contains(key)) {
            getEditor().remove(key).apply();
        }
    }

    public void deleteAll() {
        getEditor().clear().apply();
    }

    public void save(String key, Object value) {
        SharedPreferences.Editor editor = getEditor();
        if (value instanceof Boolean) {
            editor.putBoolean(key, (Boolean) value);
        } else if (value instanceof Integer) {
            editor.putInt(key, (Integer) value);
        } else if (value instanceof Float) {
            editor.putFloat(key, (Float) value);
        } else if (value instanceof Long) {
            editor.putLong(key, (Long) value);
        } else if (value instanceof String) {
            editor.putString(key, (String) value);
        } else if (value instanceof Enum) {
            editor.putString(key, value.toString());
        } else if (value != null) {
            throw new RuntimeException("Attempting to save non-supported preference");
        }

        editor.apply();
    }

    @SuppressWarnings("unchecked")
    public <T> T get(String key) {
        return (T) sharedPreferences.getAll().get(key);
    }

    @SuppressWarnings("unchecked")
    public <T> T get(String key, T defValue) {
        T returnValue = (T) sharedPreferences.getAll().get(key);
        return returnValue == null ? defValue : returnValue;
    }

    public boolean has(String key) {
        return sharedPreferences.contains(key);
    }

    private SharedPreferences.Editor getEditor() {
        return sharedPreferences.edit();
    }

    public Set<String> getSet(String key, Set<String> set) {
        return sharedPreferences.getStringSet(key, set);
    }

    public void saveSet(String key, TreeSet<String> set) {
        getEditor().putStringSet(key, set).commit();
    }
}

File: moreapps/src/main/java/io/github/raghavsatyadev/moreapps/utils/MoreAppsPrefHelper.kt
@file:Suppress("unused")

package io.github.raghavsatyadev.moreapps.utils

import android.content.Context
import android.content.SharedPreferences
import android.content.SharedPreferences.Editor
import java.util.TreeSet

internal class MoreAppsPrefHelper private constructor(context: Context) {
    private val sharedPreferences: SharedPreferences

    init {
        instance = this
        sharedPreferences = context.getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE)
    }

    fun delete(key: String?) {
        if (sharedPreferences.contains(key)) {
            editor.remove(key).apply()
        }
    }

    fun deleteAll() {
        editor.clear().apply()
    }

    fun save(key: String?, value: Any?) {
        val editor = editor
        if (value is Boolean) {
            editor.putBoolean(key, (value as Boolean?)!!)
        } else if (value is Int) {
            editor.putInt(key, (value as Int?)!!)
        } else if (value is Float) {
            editor.putFloat(key, (value as Float?)!!)
        } else if (value is Long) {
            editor.putLong(key, (value as Long?)!!)
        } else if (value is String) {
            editor.putString(key, value as String?)
        } else if (value is Enum<*>) {
            editor.putString(key, value.toString())
        } else if (value != null) {
            throw RuntimeException("Attempting to save non-supported preference")
        }
        editor.apply()
    }

    operator fun <T> get(key: String?): T? {
        return sharedPreferences.all[key] as T?
    }

    operator fun <T> get(key: String?, defValue: T): T {
        val returnValue = sharedPreferences.all[key] as T?
        return returnValue ?: defValue
    }

    fun has(key: String?): Boolean {
        return sharedPreferences.contains(key)
    }

    private val editor: Editor
        get() = sharedPreferences.edit()

    fun getSet(key: String?, set: Set<String?>?): Set<String>? {
        return sharedPreferences.getStringSet(key, set)
    }

    fun saveSet(key: String?, set: TreeSet<String?>?) {
        editor.putStringSet(key, set).commit()
    }

    companion object {
        private const val SHARED_PREFS_NAME = "MORE_APPS"

        @Volatile
        private var instance: MoreAppsPrefHelper? = null

        @Synchronized
        fun getInstance(context: Context): MoreAppsPrefHelper? {
            if (instance == null) {
                instance = MoreAppsPrefHelper(context)
            }
            return instance
        }
    }
}